addons/account_payment/controllers/portal.py

```diff
@@ -1,26 +1,54 @@
-            token_count = request.env['payment.token'].sudo().search_count([('acquirer_id.company_id', '=', invoice.company_id.id),
-                                                                      ('partner_id', '=', invoice.partner_id.id),
-                                                                    ])
-            values['existing_token'] = token_count > 0
-        values.update(payment_inputs)
-        # if the current user is connected we set partner_id to his partner otherwise we set it as the invoice partner
-        # we do this to force the creation of payment tokens to the correct partner and avoid token linked to the public user
-        values['partner_id'] = invoice.partner_id if is_public_user else request.env.user.partner_id,
+        values = super()._invoice_get_page_view_values(invoice, access_token, **kwargs)
+        logged_in = not request.env.user._is_public()
+        # We set partner_id to the partner id of the current user if logged in, otherwise we set it
+        # to the invoice partner id. We do this to ensure that payment tokens are assigned to the
+        # correct partner and to avoid linking tokens to the public user.
+        partner_id = request.env.user.partner_id.id if logged_in else invoice.partner_id.id
+        acquirers_sudo = request.env['payment.acquirer'].sudo()._get_compatible_acquirers(
+            invoice.company_id.id or request.env.company.id,
+            partner_id,
+            currency_id=invoice.currency_id.id,
+        )  # In sudo mode to read the fields of acquirers and partner (if not logged in)
+        tokens = request.env['payment.token'].search(
+            [('acquirer_id', 'in', acquirers_sudo.ids), ('partner_id', '=', partner_id)]
+        )  # Tokens are cleared at the end if the user is not logged in
+        fees_by_acquirer = {
+            acq_sudo: acq_sudo._compute_fees(
+                invoice.amount_total, invoice.currency_id, invoice.company_id.country_id
+            ) for acq_sudo in acquirers_sudo.filtered('fees_active')
+        }
+        values.update({
+            'acquirers': acquirers_sudo,
+            'tokens': tokens,
+            'fees_by_acquirer': fees_by_acquirer,
+            'show_tokenize_input': logged_in,  # Prevent public partner from saving payment methods
+            'amount': invoice.amount_residual,
+            'currency': invoice.currency_id,
+            'partner_id': partner_id,
+            'access_token': access_token,
+            'transaction_route': f'/invoice/transaction/{invoice.id}/',
+            'landing_route': _build_url_w_params(invoice.access_url, {'access_token': access_token})
+        })
+        # TODO TBE This whole behavior of fetching the logged in partner/invoice partner tokens and
+        #          then clearing them if the partner was actually logged out looks a bit odd to me
+        #          Shouldn't we do something similar to what is done in other controllers?
```

Comment:
This way, we avoid displaying payment token for non connected users, but we can warn them that existing tokens are available if you log-in. This behavior is correct IMO